rust/dev-pl011/src/lib.rs

```diff
@@ -0,0 +1,293 @@
+const UART_TFR_TXFF: u32 = 1 << 5; // UART Transmit FIFO Full
+// const UART_TFR_RXFF: u32 = 1 << 6; // UART Receive FIFO Full
+// const UART_TFR_TXFE: u32 = 1 << 7; // UART Transmit FIFO Empty
+
+const UART_IMSC_RXIM: u32 = 1 << 4; // Receive interrupt mask
+
+const UART_TMIS_RXMIS: u32 = 1 << 4; // Receive masked interrupt status
+
+const UART_CR_RXEN: u32 = 1 << 9; // Receive enable
+
+// TODO: Don't make this manually.
+#[repr(C)]
+#[derive(Clone)]
+#[allow(non_camel_case_types)]
+struct pl011_config {
+    base: usize,
+    irq: u32,
+    flag: u32,
+}
+
+/// Global state of the uart. Very _non_ rust friendly right now, just a
+/// directly translation from the C code.
+struct Pl011Uart {
+    config: pl011_config,
+    rx_buf: UnsafeCell<sys::cbuf>,
+}
+
+/// The underlying cbuf is actually Sync because it uses its own spinlock.
+unsafe impl Sync for Pl011Uart {}
+
+/// Get a reference to the given uart.
+fn get_uart(port: usize) -> &'static Pl011Uart {
+    if port >= PL011_UARTS.len() {
+        // For now, just panic if we access an invalid port.
+        panic!("pl011: invalid port {}", port);
+    }
+    PL011_UARTS[port].get()
+}
+
+/// Get a reference to the given uart that only requires early init to have been done.
+fn get_uart_early(port: usize) -> &'static Pl011Uart {
+    if port >= PL011_UARTS.len() {
+        // For now, just panic if we access an invalid port.
+        panic!("pl011: invalid port {}", port);
+    }
+    PL011_UARTS[port].get_early()
+}
+
+impl Pl011Uart {
+    /// Get the mmio handle for the given UART.
+    ///
+    /// # Safety
+    /// This routine is not marked as unsafe, as we assume that the struct has
+    /// been initialized with a proper offset of the base of the uart registers.
+    /// Although the docs for safe-mmio suggest that having multiple instances
+    /// of this handle is unsafe, this is not a rust safety issue. The hardware
+    /// block, and memory semantics are well defined for multiple concurrent
+    /// access (including writes) and incorrect use is a logic error in the
+    /// driver, not an issue of Rust memory safety.
+    ///
+    /// But, this conflation of the Rust borrow checker's rules around only one
+    /// instance of a mutable reference, and how the hardware works is
+    /// persistent. As the underlying MMIO is implemented using raw pointers,
+    /// there are no references, having multiple instances of this handle is
+    /// perfectly safe.
+    ///
+    /// It is important that the returned handle is 'static, otherwise we only
+    /// get one borrow which defeats the entire purpose of this routine.
+    fn get_dev(&self) -> UniqueMmioPointer<'static, UartRegs> {
+        unsafe { UniqueMmioPointer::new(NonNull::new(self.config.base as _).unwrap()) }
+    }
+}
+
+const RXBUF_SIZE: usize = 32;
+
+static PL011_UARTS: [LkOnce<Pl011Uart>; 1] = [LkOnce::uninit()];
+
+extern "C" fn uart_irq(arg: *mut c_void) -> sys::handler_return {
```

Comment:
super cool that this just directly lines up with the irq model.

Reply:
I do have a desire, though, to create a proc macro to make this a little easier to declare, more along the lines of something like:
```rust
#[lk::irq_handler]
fn uart_irq(arg: *mut c_void) -> sys::handler_return ...
```
but I'm not actually sure if that is all that helpful. More helpful would be something a bit more strongly typed so that the type of the handler argument would be a real rust type instead of having to see the `*mut c_void` from the rust side.

That kind of comes down to the question of how much we want to abstract the lk C mechanisms from the Rust code. And that doesn't have to be decided now, as we do more, it might become more clear what is actually wanted.